sky/backends/cloud_vm_ray_backend.py

```diff
@@ -1709,21 +1709,21 @@ def _update_stable_cluster_ips(self,
 
         def internal_ips(self,
                          max_attempts: int = 1,
-                         use_cached_ips: bool = True):
+                         use_cached_ips: bool = True) -> List[str]:
             if not use_cached_ips:
                 self._update_stable_cluster_ips(max_attempts=max_attempts)
             if self.stable_internal_external_ips is not None:
                 return [ips[0] for ips in self.stable_internal_external_ips]
-            return None
+            return []
```

Comment:
If we return an empty list, we may need to add assertions to the following places, to avoid the program failing silently, i.e. executing on empty list.
```
assert ip_list, 'ip_list should not be empty'
```
https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L2048

https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L2128
PS, the `max_attempts=_FETCH_IP_MAX_ATTEMPTS` in the line above should be removed, as it does not take any effect?

https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L2524
https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L2933
https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L2966
https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L3110

Also, we need to assert that the `internal_ips` is non-empty.
https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L3156
https://github.com/skypilot-org/skypilot/blob/5506e1dec428c8bd6dff4110ea86e6c5e76bd9dd/sky/backends/cloud_vm_ray_backend.py#L3219

Reply:
Actually, I prefer to have the return value to be None when the cache does not exist, and still have the assertion in the places above:
```
assert ip_list is not None, 'external_ips is not cached in the handle'
```

Reply:
Ok! If the return value may be `None`, then we'll also need to make sure that we never call `len`